### PR TITLE
Potential fix for pylauncher with python version

### LINKED ISSUE #1150: --python python3.11 results in an error
**Describe the bug**

- The output of `--help` explains that `pipx run` supports  `--python python3.11`.
- My testing shows that this results in an error
- I tested on 1.3.3 and the latest version from main
- A workaround is to use `--python 3.11`
- *Edit* I tested with the `python-launcher` package installed on Fedora 38

**How to reproduce**

Command to reproduce this bug:

    pipx run --python python3.11 pycowsay moo

Error message:

    Traceback (most recent call last):
      File "/home/maxwell-k/.local/bin/pipx", line 8, in <module>
        sys.exit(cli())
                 ^^^^^
      File "/home/maxwell-k/.local/pipx/venvs/pipx/lib64/python3.11/site-packages/pipx/main.py", line 885, in cli
        return run_pipx_command(parsed_pipx_args)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/maxwell-k/.local/pipx/venvs/pipx/lib64/python3.11/site-packages/pipx/main.py", line 192, in run_pipx_command
        py_launcher_python = find_py_launcher_python(args.python)
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/maxwell-k/.local/pipx/venvs/pipx/lib64/python3.11/site-packages/pipx/interpreter.py", line 33, in find_py_launcher_python
        py = subprocess.run(
             ^^^^^^^^^^^^^^^
      File "/usr/lib64/python3.11/subprocess.py", line 571, in run
        raise CalledProcessError(retcode, process.args,
    subprocess.CalledProcessError: Command '['/usr/bin/py', '-python3.11', '-c', 'import sys; print(sys.executable)']' returned non-zero exit status 2.

Command to show `pipx` version:

    pipx --version

Output:

    1.3.3

Also tested with:

    1.3.4.dev8+gb66335c

**Expected behavior**

Command to show help:

    pipx run --help

Excerpt from help:

```
  --python PYTHON       Python to run with. Possible values can be the executable
                        name (python3.11), the version to pass to py launcher (3.11),
                        or the full path to the executable. Requires Python 3.8 or
                        above.
```

I therefore expect the two commands below to be equivalent:

    pipx run --python python3.11 pycowsay moo
    pipx run --python 3.11 pycowsay moo

They should both display a mooing cow:

      ---
    < moo >
      ---
       \   ^__^
        \  (oo)\_______
           (__)\       )\/\
               ||----w |
               ||     ||

### PR DESCRIPTION
**Pull request by Gitznik**
- [X] I have added an entry to `docs/changelog.md`

## Summary of changes

When pylauncher is detected, check the provided python version to see if it was provided including `python` (as per our examples). If so, log a warning and remove python from the version, leaving only the semver, as expected by pylauncher.

This might fix #1150

## Test plan

Not tested yet, as I'm not on windows and I don't have pylauncher. I've added an automated test that might verify the new behavior. 
Maybe someone using pylauncher can assist me here with a manual validation.

```
# command(s) to exercise these changes
```

### CODE REVIEW

**Comment by maxwell-k** [comment]
TLDR; in my opinion, this change is good and doesn't fix  #1150 

Testing on Linux, this change causes expected behaviour more often and prints a helpful warning message. The change also means that if a version of Python is not installed, all three ways of specifying that version (e.g. (1) `3.11`, (2) `python3.11`, (3) `/usr/bin/python3.11`) produce the same traceback.

This testing also illustrates https://github.com/pypa/pipx/issues/1150 effectively; without the python launcher installed and using either the main branch or this one the results include five different trace backs. I think my expectation was that these would all be an error message. I appreciate that would be a separate PR.

---



Testing the main branch for a baseline understanding:

- with Python 3.12 installed
- without Python 3.11 installed
- without python launcher installed

Summary

    pipx run pycowsay moo  # expected output
    pipx run --python=/usr/bin/python3.12 pycowsay moo  # expected output
    pipx run --python=python3.12 pycowsay moo  # trace back 1
    pipx run --python=3.12 pycowsay moo  # trace back 2
    pipx run --python=/usr/bin/python3.11 pycowsay moo  # trace back 3
    pipx run --python=python3.11 pycowsay moo  # trace back 4
    pipx run --python=3.11 pycowsay moo  # trace back 5



Testing the `pylauncher_fix_version` branch:

- with Python 3.12 installed
- without Python 3.11 installed
- without python launcher installed

The summary is as above.



Testing the `pylauncher_fix_version` branch:

- with Python 3.12 installed
- without Python 3.11 installed
- **with** python launcher installed

Summary:

    pipx run pycowsay moo  # expected output
    pipx run --python=/usr/bin/python3.12 pycowsay moo  # expected output
    pipx run --python=python3.12 pycowsay moo # warning then expected output
    pipx run --python=3.12 pycowsay moo  # expected output
    pipx run --python=/usr/bin/python3.11 pycowsay moo  # trace back 6
    pipx run --python=python3.11 pycowsay moo  # trace back 6
    pipx run --python=3.11 pycowsay moo  # trace back 6

<details>
<summary>Expected output</summary>

```
  ---
< moo >
  ---
   \   ^__^
    \  (oo)\_______
       (__)\       )\/\
           ||----w |
           ||     ||

```

</details>

<details>
<summary>Trace back 1</summary>

```
Traceback (most recent call last):
  File "/home/maxwell-k/.local/bin/pipx", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/main.py", line 914, in cli
    return run_pipx_command(parsed_pipx_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/main.py", line 197, in run_pipx_command
    commands.run(
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/commands/run.py", line 200, in run
    run_package(
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/commands/run.py", line 157, in run_package
    _download_and_run(
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/commands/run.py", line 234, in _download_and_run
    venv.create_venv(venv_args, pip_args, override_shared)
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/venv.py", line 160, in create_venv
    venv_process = run_subprocess(cmd + venv_args + [str(self.root)])
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/util.py", line 179, in run_subprocess
    completed_process = subprocess.run(
                        ^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/subprocess.py", line 548, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib64/python3.12/subprocess.py", line 1950, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/home/maxwell-k/python3.12'

```

</details>

<details>
<summary>Trace back 2</summary>

```
Traceback (most recent call last):
  File "/home/maxwell-k/.local/bin/pipx", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/main.py", line 914, in cli
    return run_pipx_command(parsed_pipx_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/main.py", line 197, in run_pipx_command
    commands.run(
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/commands/run.py", line 200, in run
    run_package(
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/commands/run.py", line 157, in run_package
    _download_and_run(
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/commands/run.py", line 234, in _download_and_run
    venv.create_venv(venv_args, pip_args, override_shared)
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/venv.py", line 160, in create_venv
    venv_process = run_subprocess(cmd + venv_args + [str(self.root)])
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/util.py", line 179, in run_subprocess
    completed_process = subprocess.run(
                        ^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/subprocess.py", line 548, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib64/python3.12/subprocess.py", line 1950, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/home/maxwell-k/3.12'
```

</details>

<details>
<summary>Trace back 3</summary>

```
Traceback (most recent call last):
  File "/home/maxwell-k/.local/bin/pipx", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/main.py", line 914, in cli
    return run_pipx_command(parsed_pipx_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/main.py", line 197, in run_pipx_command
    commands.run(
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/commands/run.py", line 200, in run
    run_package(
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/commands/run.py", line 157, in run_package
    _download_and_run(
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/commands/run.py", line 234, in _download_and_run
    venv.create_venv(venv_args, pip_args, override_shared)
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/venv.py", line 160, in create_venv
    venv_process = run_subprocess(cmd + venv_args + [str(self.root)])
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/util.py", line 179, in run_subprocess
    completed_process = subprocess.run(
                        ^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/subprocess.py", line 548, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib64/python3.12/subprocess.py", line 1950, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/usr/bin/python3.11'
```

</details>

<details>
<summary>Trace back 4</summary>

```
Traceback (most recent call last):
  File "/home/maxwell-k/.local/bin/pipx", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/main.py", line 914, in cli
    return run_pipx_command(parsed_pipx_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/main.py", line 197, in run_pipx_command
    commands.run(
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/commands/run.py", line 200, in run
    run_package(
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/commands/run.py", line 157, in run_package
    _download_and_run(
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/commands/run.py", line 234, in _download_and_run
    venv.create_venv(venv_args, pip_args, override_shared)
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/venv.py", line 160, in create_venv
    venv_process = run_subprocess(cmd + venv_args + [str(self.root)])
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/util.py", line 179, in run_subprocess
    completed_process = subprocess.run(
                        ^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/subprocess.py", line 548, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib64/python3.12/subprocess.py", line 1950, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/home/maxwell-k/python3.11'
```

</details>

<details>
<summary>Trace back 5</summary>

```
Traceback (most recent call last):
  File "/home/maxwell-k/.local/bin/pipx", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/main.py", line 914, in cli
    return run_pipx_command(parsed_pipx_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/main.py", line 197, in run_pipx_command
    commands.run(
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/commands/run.py", line 200, in run
    run_package(
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/commands/run.py", line 157, in run_package
    _download_and_run(
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/commands/run.py", line 234, in _download_and_run
    venv.create_venv(venv_args, pip_args, override_shared)
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/venv.py", line 160, in create_venv
    venv_process = run_subprocess(cmd + venv_args + [str(self.root)])
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/util.py", line 179, in run_subprocess
    completed_process = subprocess.run(
                        ^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/subprocess.py", line 548, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib64/python3.12/subprocess.py", line 1950, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/home/maxwell-k/3.11'
```

</details>

<details>
<summary>Trace back 6</summary>

```
Traceback (most recent call last):
  File "/home/maxwell-k/.local/bin/pipx", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/main.py", line 911, in cli
    return run_pipx_command(parsed_pipx_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/main.py", line 192, in run_pipx_command
    py_launcher_python = find_py_launcher_python(args.python)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maxwell-k/.local/lib/python3.12/site-packages/pipx/interpreter.py", line 33, in find_py_launcher_python
    py = subprocess.run(
         ^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/usr/bin/py', '-3.11', '-c', 'import sys; print(sys.executable)']' returned non-zero exit status 64.


```

</details>

**Comment by Gitznik** [comment]
I'll convert this to draft for now as it only fixes a subset of the mentioned problems.

**Comment by Gitznik** [comment]
> TLDR; in my opinion, this change is good ~and doesn't fix #1150~
> 
> Testing on Linux, this change causes expected behaviour more often and prints a helpful warning message. The change also means that if a version of Python is not installed, all three ways of specifying that version (e.g. (1) `3.11`, (2) `python3.11`, (3) `/usr/bin/python3.11`) produce the same traceback.
> 
> This testing also illustrates another part of #1150 ; without the python launcher installed and using either the main branch or this one the results include five different trace backs. It can be difficult to see how these relate to the command that triggered them. I think my expectation was that these would print a message from which I could understand I need to install the launcher. I appreciate that would be a separate PR to this one.
> 
> Testing the main branch for a baseline understanding:
> 
> * with Python 3.12 installed
> * without Python 3.11 installed
> * without python launcher installed
> 
> Summary
> 
> ```
> pipx run pycowsay moo  # expected output
> pipx run --python=/usr/bin/python3.12 pycowsay moo  # expected output
> pipx run --python=python3.12 pycowsay moo  # trace back 1
> pipx run --python=3.12 pycowsay moo  # trace back 2
> pipx run --python=/usr/bin/python3.11 pycowsay moo  # trace back 3
> pipx run --python=python3.11 pycowsay moo  # trace back 4
> pipx run --python=3.11 pycowsay moo  # trace back 5
> ```
> 
> Testing the `pylauncher_fix_version` branch:
> 
> * with Python 3.12 installed
> * without Python 3.11 installed
> * without python launcher installed
> 
> The summary is as above.
> 
> Testing the `pylauncher_fix_version` branch:
> 
> * with Python 3.12 installed
> * without Python 3.11 installed
> * **with** python launcher installed
> 
> Summary:
> 
> ```
> pipx run pycowsay moo  # expected output
> pipx run --python=/usr/bin/python3.12 pycowsay moo  # expected output
> pipx run --python=python3.12 pycowsay moo # warning then expected output
> pipx run --python=3.12 pycowsay moo  # expected output
> pipx run --python=/usr/bin/python3.11 pycowsay moo  # trace back 6
> pipx run --python=python3.11 pycowsay moo  # trace back 6
> pipx run --python=3.11 pycowsay moo  # trace back 6
> ```
> 
> Expected output
> Trace back 1
> Trace back 2
> Trace back 3
> Trace back 4
> Trace back 5
> Trace back 6

I've added some changes to unify how the python interpreter is evaluated based on the --python flag. Still need to clean up the now redundant checks and add some tests, but looks promising IMO.

Testing the pylauncher_fix_version branch:

No Pylauncher, 3.11 installed, 3.10 not installed

pipx run pycowsay moo  # expected output
pipx run --python=/home/robert/.pyenv/shims/python3.11 pycowsay moo  # expected output
pipx run --python=python3.11 pycowsay moo  # expected output
pipx run --python=3.11 pycowsay moo  # warning1
pipx run --python=/home/robert/.pyenv/shims/python3.10 pycowsay moo  # warning2
pipx run --python=python3.10 pycowsay moo  # warning2
pipx run --python=3.10 pycowsay moo  # warning1

Pylauncher, 3.11 installed, 3.10 not installed

pipx run pycowsay moo  # expected output
pipx run --python=/usr/bin/python3.10 pycowsay moo  # expected output
pipx run --python=python3.11 pycowsay moo  # expected output
pipx run --python=3.11 pycowsay moo  # expected output
pipx run --python=/home/robert/.pyenv/shims/python3.10 pycowsay moo  # warning3, this could be improved maybe
pipx run --python=python3.10 pycowsay moo  # warning3
pipx run --python=3.10 pycowsay moo  # warning3

warning1:
⚠️ No executable for the provided Python version '3.XX' found in PATH.
    Please make sure the provided version is on your PATH or the file
    path is valid. The provided version looks like a version for Python
    Launcher, but `py` was not found on PATH.

warning2:
⚠️ No executable for the provided Python version
    '<python version/path>' found in PATH. Please make
    sure the provided version is on your PATH or the file path is valid.

warning3:
⚠️ No executable for the provided Python version '<python version/path>'
    found in py launcher. Please make sure the provided version is listed
    when running `py --list`.

File: src/pipx/commands/install.py [hunk 1]
```diff
@@ -107,23 +106,6 @@ def install(
...
-                           Please make sure the executable name is on your PATH /
-                           the path to the executable is correct.
-                        """,
-                        subsequent_indent=" " * 4,
-                    )
-                )
-                return EXIT_CODE_SPECIFIED_PYTHON_EXECUTABLE_NOT_FOUND
-            else:
-                print()
-                raise
```
Comment (Gitznik) [inline src/pipx/commands/install.py:126]: I have removed this block as it's now covered by the new check at the start of the application. I implemented it in https://github.com/pypa/pipx/pull/1171

File: src/pipx/interpreter.py [hunk 2]
```diff
@@ -17,6 +21,46 @@ def has_venv() -> bool:
...
+
+    try:
+        py_executable = find_py_launcher_python(python_version)
+        if py_executable:
+            return py_executable
+    except (subprocess.CalledProcessError, FileNotFoundError) as e:
+        raise InterpreterResolutionError(source="py launcher", version=python_version) from e
+
+    if shutil.which(python_version):
+        return python_version
```
Comment (guahki) [inline src/pipx/interpreter.py:60]: Hi @Gitznik,
after a little looking into it, I would like to see ``return shutil.which(python_version)`` here.

The reasoning is: I want to be able to give a python executable residing in the path on Windows with e.g. ``--python3.12.exe``. However, since #1168 this is no longer possible, as ``os.path.realpath`` would prepend the cwd and thus, it tries to run ``<current_path>\python3.12.exe``, which doesn't exist, but prevents ``subprocess.run`` from looking up the python executable on the path. (Sidenote: a similar problem was resolved in #1196 for non-Windows platform).

This PR does not change the current situation, however, if the above change would be introduced, the problematic code introduced in #1168 would get a fully resolved path (from ``shutil.which``) and can then perform ``os.,path.realpath`` on that without breaking non-absolute paths.

It remains to say: I personally think, #1168 in this form is a mistake and should never have been merged. Or at least only with much more restrictions, on when ``os.path.realpath`` is executed (e.g. only for abolute paths).

Comment (Gitznik) [reply src/pipx/interpreter.py:60]: Hi guahki,

thanks for your suggestion. While I think it makes sense, I would suggest creating a separate issue for the bug you're describing. Changing the provided python version to a path may or may not have other implications that should be discussed and tested separately.

Comment (guahki) [reply src/pipx/interpreter.py:60]: Did that in #1205. Thanks for the quick answer

**Review by gaborbernat — CHANGES_REQUESTED** [review]
Let's rebase and add changelog.

**Review by gaborbernat — APPROVED** [review]
